### PR TITLE
Prevent get_tops from performing a Set operation on a List

### LINKED ISSUE #61176: [BUG] Rendering Primary Top file failed when pillar_source_merging_strategy: none
**Description**
state.\*, pillar.\* etc fail with
```
root@node1 ~ # salt-call pillar.items
local:
    ----------
    _errors:
        - Rendering Primary Top file failed, render error:
          unsupported operand type(s) for &=: 'list' and 'set'
```

**Setup**
```
file_roots:
  alias1:
    - /root/salt4/salt/alias1
    - /root/salt4/salt/base
  alias2:
    - /root/salt4/salt/alias2
    - /root/salt4/salt/base
pillar_source_merging_strategy: none
```
  
**Steps to Reproduce the behavior**

call `salt-call pillar.items` or `salt-call state.apply` on master or any minion.

```
2021-11-03 12:15:26,611 [salt.pillar      :691 ][ERROR   ][874238] Pillar rendering failed for minion saltmaster
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/pillar/__init__.py", line 671, in get_tops
    saltenvs &= {self.saltenv or "base"}
TypeError: unsupported operand type(s) for &=: 'list' and 'set'
2021-11-03 12:15:26,611 [salt.pillar      :1220][CRITICAL][874238] Pillar render error: Rendering Primary Top file failed, render error:
unsupported operand type(s) for &=: 'list' and 'set'
2021-11-03 12:15:26,778 [salt.pillar      :691 ][ERROR   ][874240] Pillar rendering failed for minion saltmaster
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/pillar/__init__.py", line 671, in get_tops
    saltenvs &= {self.saltenv or "base"}
TypeError: unsupported operand type(s) for &=: 'list' and 'set'
2021-11-03 12:15:26,779 [salt.pillar      :1220][CRITICAL][874240] Pillar render error: Rendering Primary Top file failed, render error:
unsupported operand type(s) for &=: 'list' and 'set'
```

**Versions Report**

<details>
Salt Version:
          Salt: 3004
 
Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: 1.4.4
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.6.1
        pygit2: Not Installed
        Python: 3.8.10 (default, Sep 28 2021, 16:10:42)
  python-gnupg: 0.4.5
        PyYAML: 5.3.1
         PyZMQ: 18.1.1
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.2
 
System Versions:
          dist: ubuntu 20.04 focal
        locale: utf-8
       machine: x86_64
       release: 5.4.0-89-generic
        system: Linux
       version: Ubuntu 20.04 focal
</details>

**Additional context**

```/usr/lib/python3/dist-packages/salt/pillar/__init__.py```

```
        try:
            saltenvs = set()
            if self.opts["pillarenv"]:
                # If the specified pillarenv is not present in the available
                # pillar environments, do not cache the pillar top file.
                if self.opts["pillarenv"] not in self.opts["pillar_roots"]:
                    log.debug(
                        "pillarenv '%s' not found in the configured pillar "
                        "environments (%s)",
                        self.opts["pillarenv"],
                        ", ".join(self.opts["pillar_roots"]),
                    )
                else:
                    saltenvs.add(self.opts["pillarenv"])
            else:
                saltenvs = self._get_envs()
                if self.opts.get("pillar_source_merging_strategy", None) == "none":
                    saltenvs &= {self.saltenv or "base"}
```

### PR DESCRIPTION
**Pull request by binocvlar**
### What does this PR do?

Prior to this commit, `get_tops` attempted to perform a Set
intersection_update (`&=`) using the List returned by the `_get_envs`
method as the right-hand operand (this raises `TypeError`). When this
code was evaluated, it would result in the following error:

```
Rendering Primary Top file failed, render error: unsupported operand
type(s) for &=: 'list' and 'set'
```

Unpack the list into a new Set to permit the intersection update to
succeed.

See https://github.com/saltstack/salt/issues/61176

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/61176

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

No tests have been written here, however, this change seems locally constrained, and simply fixes a `TypeError`.

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/pillar/__init__.py
Comment (waynew) [inline salt/pillar/__init__.py:695]: **question** would `saltenvs.update(self._get_envs())` be more accurate here?

**Review by waynew — CHANGES_REQUESTED** [review]
Thanks for the PR! This looks like it should fix the problem... We are requiring tests before we can merge PRs. If you need help writing tests to cover this, attend one of the [Test Clinics on Twitch](https://www.twitch.tv/saltstackinc/schedule), and I'd be happy to point you in the right direction :+1:

**Comment by waynew** [comment]
I had some time so I went ahead and ended out writing a test case for this on today's Test Clinic:

```
def test_pillar_get_tops_should_not_error_when_merging_strategy_is_none_and_no_pillarenv(
    temp_salt_minion,
):
    opts = temp_salt_minion.config.copy()
    opts["pillarenv"] = None
    opts["pillar_source_merging_strategy"] = "none"
    pillar = salt.pillar.Pillar(
        opts=opts,
        grains=salt.loader.grains(opts),
        minion_id=temp_salt_minion.id,
        saltenv="base",
    )
    tops, errors = pillar.get_tops()
    assert not errors
```

That belongs in https://github.com/saltstack/salt/blob/master/tests/pytests/unit/pillar/test_pillar.py.

This PR will also need a [changelog entry](https://docs.saltstack.com/en/latest/topics/development/changelog.html), and then it will be ready for merge :+1:

File: salt/pillar/__init__.py
Comment (binocvlar) [reply salt/pillar/__init__.py:695]: Hi @waynew : it's _probably_ more in keeping with the style of this function (given that the Set assigned to `saltenvs` on L681 is clearly destined for mutation). With that said, the line that I replaced was not attempting to mutate `saltenvs`, but rather, rebound the variable name to a new set (well, actually a list, due to the reported bug). This, combined with the fact that I usually attempt to avoid mutation led me to rebind it as well.

I'm happy to update the PR if your preference is to mutate - just let me know :)

**Review by Ch3LL — CHANGES_REQUESTED** [review]
This PR just needs a test added. Can you copy/paste the test @waynew provided into the file he defined in an earlier comment?

**Comment by aarnaud** [comment]
This PR may fix a regression in 3004:
```
    _errors:
        - Rendering Primary Top file failed, render error:
          unsupported operand type(s) for &=: 'list' and 'set'
```
Not present in 3003

So until there is a fix in 3004, I had to rollback to 3003.

**Comment by binocvlar** [comment]
> This PR just needs a test added. Can you copy/paste the test @waynew provided into the file he defined in an earlier comment?

Thanks for the nudge @Ch3LL - that's been done now :)

**Review by Ch3LL — DISMISSED** [review]

File: salt/pillar/__init__.py
Comment (waynew) [reply salt/pillar/__init__.py:695]: I'm +0 on it :joy: 

The Boy Scout in me says that we should go ahead and make the change, so that it isn't confusing anyone down the road.

The pragmatist in me says :shrug: 

I guess... if you're happy enough to make the change, then let's make the change and get this merged in :+1:

Comment (waynew) [reply salt/pillar/__init__.py:695]: FWIW, I went through the git history for this function and... it definitely had some changes. IIRC the *desired* behavior actually was to do some mutation

Comment (binocvlar) [reply salt/pillar/__init__.py]: No worries @waynew - I've just pushed that change now :)

**Review by waynew — APPROVED** [review]
:rocket:

**Review by Ch3LL — APPROVED** [review]

**Comment by vineet-kum** [comment]
this is still not part of the latest release [3004.1 ](https://github.com/saltstack/salt/releases/tag/v3004.1):(

**Comment by vin01** [comment]
this is still not part of the latest release [3004.1 ](https://github.com/saltstack/salt/releases/tag/v3004.1):(

**Comment by lainatnavi** [comment]
I'm having this issue after updating salt-master form 3004 to 3004.1.

**Comment by mugurelu** [comment]
Still not fixed in 3004.2 :(

**Comment by mindflayer** [comment]
Any chance to see this bugfix in a 3004.x release?

**Comment by Ch3LL** [comment]
This will be included in the 3005 release. There are no current plans to add it to a 3004 point release.

**Comment by mugurelu** [comment]
What is the logic behind the decision to not have it in 3004? So if you have a bug in one major version, it stays there for half a year until the next major version? In this case, what’s released in the minor versions?

**Comment by vin01** [comment]
This really needs to get in a bug fix release, relying on automatic upgrades to make sure latest patches are applied is a reasonable thing to do, and in such cases with bug fixes with such a broad impact waiting for months until making it to a release, they break the manually patched installations *every* single time.

**Comment by mindflayer** [comment]
> This will be included in the 3005 release. There are no current plans to add it to a 3004 point release.

I don't really understand this decision, it's an issue hitting several people, and the fix is a one-line patch that I was able to manually apply.

**Pull request merged**
